File: GW3/GWLogger/Backend/DataContext/Query/Tokenizer.cs
using GWLogger.Backend.DataContext.Query.Tokens;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query
{
    class Tokenizer
    {
        List<Token> Tokens { get; } = new List<Token>();
        public int Position { get; set; }
        static List<Token> KnownTokenTypes;

        static Tokenizer()
        {
            KnownTokenTypes = Assembly.GetExecutingAssembly()
                .DefinedTypes
                .Where(row => row.IsSubclassOf(typeof(Token)))
                .Select(row => (Token)row.GetConstructor(new Type[] { })
                    .Invoke(new object[] { })).ToList();
        }

        internal Tokenizer(QueryParser parser)
        {
            while (parser.HasChar())
            {
                parser.SkipSpaces();
                var possibleToken = KnownTokenTypes.FirstOrDefault(row => row.CanBeUsed(parser));
                if (possibleToken == null)
                    throw new InvalidTokenException("Token '" + parser.PeekChar() + "' not expected.");
                Tokens.Add(possibleToken.Extract(parser));
            }
        }

        internal bool HasToken()
        {
            return Position < this.Tokens.Count;
        }

        public Token Peek(int offset = 0)
        {
            return Tokens[Position + offset];
        }

        internal Token Next()
        {
            return Tokens[Position++];
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokenizer.ts
class Tokenizer {
    public tokens: Token[] = [];
    public position: number = 0;
    public knownTokens: Token[] = [
        new TokenAnd(),
        new TokenAscending(),
        new TokenComa(),
        new TokenCompare(),
        new TokenOperation(),
        new TokenDescending(),
        new TokenGroup(),
        new TokenLimit(),
        new TokenSelect(),
        new TokenOr(),
        new TokenOrder(),
        new TokenStar(),
        new TokenWhere(),
        new TokenString(),
        new TokenName(),
        new TokenNumber(),
        new TokenOpenParenthesis(),
        new TokenCloseParenthesis()
    ];

    constructor(parser: QueryParser) {
        while (parser.hasChar()) {
            parser.skipSpaces();
            var possibleToken = this.knownTokens.filter(token => token.canBeUsed(parser))[0];
            if (typeof (possibleToken) != 'undefined' && possibleToken != null) {
                this.tokens.push(possibleToken.extract(parser));
            } else {
                if (typeof (parser.peekChar()) != "undefined") {
                    this.tokens.push(new TokenEnd(parser.peekString()));
                } else {
                    this.tokens.push(new TokenEnd());
                }
                break;
            }
        }
    }

    public hasToken(): boolean {
        return this.position < this.tokens.length;
    }

    public next(): Token {
        return this.tokens[this.position++];
    }

    public peek(offset: number = 0): Token {
        return this.tokens[this.position + offset];
    }
}